### DISCUSSION TITLE
How do I know if a tasklist is overloaded

### DISCUSSION THREAD

**Discussion by zhangkyou** [uber/cadence#4403] [Q&A] [answered]
How can I realized the tasklist backlog much more task, while the workers polls task from it too slowly?

Are there some metrics can helping this?

**Comment by longquanzheng** [comment, marked as answer]
There is two metric of scheduled to start latency for both decision and activity task: https://github.com/uber-go/cadence-client/blob/b09692f6838f3c45fe02fde689787c87d925bdaf/internal/common/metrics/constants.go#L46

if you are sure that workers are enough and idle, than the only possibility of high latency is tasklist being overloaded

**Reply by zhangkyou** [reply]
I find this metric is reported by worker.
How can I send metric with tally In worker code?
There is not any config to specify the tally's reporter address.

**Reply by zhangkyou** [reply]
I find the cadence-0.20.0/common/service/config/metrics.go file in cadence-server.It contains a NewScope function, this function set server's tally's scope,but why this feature dont include in cadence-client?

**Reply by longquanzheng** [reply]
Check this docs https://cadenceworkflow.io/docs/operation-guide/monitor/

**Reply by zhangkyou** [reply]
I red this, this is a Prometheus metrics, I implement a statsd metrics by my self. Another question is that what is the different between WorkerMetricScope and ServiceMetricScope? Does the cadence server's metrics can emit in worker machine too?

**Reply by longquanzheng** [reply]
Server and client metrics are separate. The two scopes you implemented are both client side metrics. WorkerScope is for worker operation metrics, like polling and processing activity and decision tasks. ServiceScope is for non worker , like starting workflow

**Comment by zhangkyou** [comment]
In this pieces of cadence client code
``` golang
func (atp *activityTaskPoller) pollWithMetrics(ctx context.Context,
	pollFunc func(ctx context.Context) (*s.PollForActivityTaskResponse, time.Time, error)) (interface{}, error) {

	response, startTime, err := pollFunc(ctx)
	if err != nil {
		return nil, err
	}
	if response == nil || len(response.TaskToken) == 0 {
		return &activityTask{}, nil
	}

	workflowType := response.WorkflowType.GetName()
	activityType := response.ActivityType.GetName()
	metricsScope := getMetricsScopeForActivity(atp.metricsScope, workflowType, activityType)
	metricsScope.Counter(metrics.ActivityPollSucceedCounter).Inc(1)
	metricsScope.Timer(metrics.ActivityPollLatency).Record(time.Now().Sub(startTime))

	scheduledToStartLatency := time.Duration(response.GetStartedTimestamp() - response.GetScheduledTimestampOfThisAttempt())
	metricsScope.Timer(metrics.ActivityScheduledToStartLatency).Record(scheduledToStartLatency)

	return &activityTask{task: response, pollStartTime: startTime}, nil
}
```

the metrics scope is atp's(*activityTaskPoller) scope, is this atp's scope derived from the worker.Options's metrics scope?
```golang
workerOptions := worker.Options{
		MetricsScope:                     h.WorkerMetricScope,
		Logger:                           h.Logger,
		MaxConcurrentActivityTaskPollers: 10,
		MaxConcurrentDecisionTaskPollers: 10,
	}
	h.StartWorkers(domain, taskListName, workerOptions)
```

**Reply by zhangkyou** [reply]
I already solved it.Please ignore this question.

**Comment by zhangkyou** [comment]
I had catch the worker's metrics. The metrics 'cadence-activity-poll-transient-failed' and 'cadence-decision-poll-transient-failed' counter is very high? Why the poll always failed?

**Reply by longquanzheng** [reply]
This means server is probably overloaded: 
https://github.com/uber-go/cadence-client/blob/999748dad95b467b7a869f10810069adfb0df643/internal/internal_retry.go#L64

Transient errors are errors like busy(ratelimiting)/timeout etc. You can try add more nodes to the server cluster.

**Reply by zhangkyou** [reply]
I only have 10 workerflows in running, several tasklists,and each frontend/matching/history/worker  service deploy in one machine.Is this throughput the one server can not cover?

**Reply by longquanzheng** [reply]
It depends on the machine perf. Check cpu/memory and network capacity

**Reply by zhangkyou** [reply]
I check the metrics on the server side.There is a metric named 'cadence_client_requests', the tag with this metric cadence_role='matching_client',cadence_service=('cadence_frontend' and 'cadence_matching') has almostly 150 every second, is this normal with an only has 10+ workflows cluster?

**Reply by longquanzheng** [reply]
It depends on the workflow code and yes, that’s possible. 
You can also look at the api request breakdown to know which apis are called the most

**Reply by zhangkyou** [reply]
I accomplished my worker's metrics emiting. Some latency metrics are very high.
'cadence-activity-poll-latency' and 'cadence-decision-poll-latency' both mostly reached 60s.
 'cadence-PollForActivityTask' and 'cadence-PollForDecisionTask' with cadence-latency tag also mostly reached 60s.
Other latency metrics like 'cadence-activity-execution-latency' and 'cadence-activity-scheduled-to-start-latency' are very low.
Is there anything wrong with poller?

**Reply by longquanzheng** [reply]
No, the high latency is expected for poll API because it’s doing long polling. Also GetWorkflowHistory also have long polling options (which is confusing and we are working separate it ) .

**Comment by zhangkyou** [comment]
another scene is I have this stack trace of the workflow that Timeout START_TO_CLOSE in my cadence-web UI.
```
coroutine 1 [blocked on chan-6.Receive]:
go.uber.org/cadence/internal.(*decodeFutureImpl).Get(0xc0014a1b20, 0x16710a0, 0xc000faf800, 0x0, 0x0, 0x2, 0x2)
	/root/go/pkg/mod/go.uber.org/cadence@v0.17.1-0.20210609205819-61495d91ff9d/internal/internal_workflow.go:1281 +0x69
git.xiaojukeji.com/falcon-server/cadence-worker/logic.SerialRawSqlWorkflow(0x16710a0, 0xc000faf800, 0xc000ce9e00, 0x0, 0x0)
	/gaia/workspace-job/git.xiaojukeji.com/falcon-server/cadence-worker/logic/cronWorkflow.go:211 +0xa59
reflect.Value.call(0x12239a0, 0x14a84a8, 0x13, 0x142f412, 0x4, 0xc000faf6e0, 0x2, 0x2, 0x2, 0x18, ...)
	/usr/local/go1.13.3/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0x12239a0, 0x14a84a8, 0x13, 0xc000faf6e0, 0x2, 0x2, 0x1, 0x2, 0xc000faf2c0)
	/usr/local/go1.13.3/src/reflect/value.go:321 +0xb4
go.uber.org/cadence/internal.(*workflowEnvironmentInterceptor).ExecuteWorkflow(0xc000faf2c0, 0x1670e60, 0xc000aca500, 0xc000b1a960, 0x4a, 0xc0009cb960, 0x1, 0x1, 0x1, 0x1, ...)
	/root/go/pkg/mod/go.uber.org/cadence@v0.17.1-0.20210609205819-61495d91ff9d/internal/workflow.go:372 +0x2b2
go.uber.org/cadence/internal.(*workflowExecutor).Execute(0xc0014a0be0, 0x1670e60, 0xc000aca500, 0xc0002a0280, 0x253, 0x253, 0xc00082b701, 0x430101, 0xc00082b6f0, 0xcbcb07, ...)
	/root/go/pkg/mod/go.uber.org/cadence@v0.17.1-0.20210609205819-61495d91ff9d/internal/internal_worker.go:699 +0x295
go.uber.org/cadence/internal.(*syncWorkflowDefinition).Execute.func1(0x16710a0, 0xc000faf5c0)
	/root/go/pkg/mod/go.uber.org/cadence@v0.17.1-0.20210609205819-61495d91ff9d/internal/internal_workflow.go:466 +0x105
```

the code in cadence-worker/logic/cronWorkflow.go:211 is only a activity execute:
```golang
	err = workflow.ExecuteActivity(ctx, SendJsonToDDMQV1Activity, featureValue, params).Get(ctx, nil)
```
why it pending in the Get funciton, I dont need the return value of this activity.

**Reply by longquanzheng** [reply]
It’s pending on Get. Get is a blocking call to wait for activity results

**Reply by longquanzheng** [reply]
If you don’t need to wait, then remove the Get. ExecuteActivity will return a future and it’s not blocking
